### PR TITLE
Add a new interface for getRepoId

### PR DESCRIPTION
**Pull request by catto**
I added `getId` base function for screwdriver-cd/screwdriver#160 .
An actual process will be implemented in scm-github.

### CODE REVIEW

**Comment by stjohnjohnson** [comment]
Nice, @catto!  I'm thinking this probably needs three more changes:

1. Change to `getRepoId`
2. Use input/output validation (suggestion on parent ticket)
3. Return a complete structure (suggestion on parent ticket)

**Comment by catto** [comment]
Thank you @stjohnjohnson! I'll work on it.
Should I add the suggested schema  to `data-schema` or wait for your push?

**Comment by stjohnjohnson** [comment]
I should have one coming this morning.

**Comment by catto** [comment]
I changed the function name to `getRepoId` and added validations for input/output.

**Comment by d2lam** [comment]
👍  LGTM 
morning-monday-hqnMFgrNktTBS

File: index.js [hunk 1]
```diff
@@ -131,6 +131,36 @@ class ScmBase {
...
+     * @param  {Object}    config        Configuration
+     * @param  {String}    config.scmUrl The scmUrl to generate ID
+     * @param  {String}    config.token  The token used to authenticate to the SCM
+     * @return {Promise}
+     */
+    getRepoId(config) {
+        const result = Joi.validate(config, dataSchema.plugins.scm.getRepoId);
+
+        if (result.error) {
+            return Promise.reject(result.error);
```
Comment (petey) [inline index.js]: I'm seeing a pattern a lot in this code, e.g. Check validation and reject on failure. Seems like a really good thing to refactor so these function's promise chains are easier to understand.

e.g.
```
validate(obj, schema) {
    const result = Joi.validate(obj, schema);
     
    if (result.error) {
         return Promise.reject(result.error);
    }
    
    return Promise.resolve(obj);
}
```

So you'd end up with something much less complex:
```
return validate(config, dataSchema.plugins.scm.getRepoId)
    .then(config => this._getRepoId(config))
    .then(repo => validate(repo, dataSchema.core.scm.repo);
```

File: index.js [hunk 2]
```diff
@@ -131,6 +131,36 @@ class ScmBase {
...
+    /**
+     * Return a unique identifier for the scmUrl
+     * @method getRepoId
+     * @param  {Object}    config        Configuration
+     * @param  {String}    config.scmUrl The scmUrl to generate ID
+     * @param  {String}    config.token  The token used to authenticate to the SCM
+     * @return {Promise}
+     */
+    getRepoId(config) {
+        const result = Joi.validate(config, dataSchema.plugins.scm.getRepoId);
```
Comment (petey) [inline index.js]: This code is going through the effort of validating the `config` input, but doesn't seem to be using it anywhere. Should it be passed to `_getRepoId`?

Comment (d2lam) [inline index.js]: `getRepoId` is what will be called first to do validation. The actual implementation is inside `_getRepoId`, which is implemented in https://github.com/screwdriver-cd/scm-github/pull/8.

Comment (d2lam) [inline index.js]: Oh I misread your comment. Yeah it should be `this._getRepoId(config)`

File: index.js (hunk 1 above)
Comment (catto) [inline index.js]: I have changed to do that!

**Comment by petey** [comment]
I think this is fantastic. 👍  Thanks @catto !

200w

**Comment by catto** [comment]
thanks!

**Pull request merged**